pages/utilities/hash-generator.tsx

```diff
@@ -0,0 +1,274 @@
+import CallToActionGrid from "@/components/CallToActionGrid";
+import Meta from "@/components/Meta";
+import { CMDK } from "@/components/CMDK";
+import { Combobox } from "@/components/ds/ComboboxComponent";
+import { Input } from "@/components/ds/InputComponent";
+import crypto, { BinaryToTextEncoding } from "crypto";
+import { generateHash } from "@/components/utils/hash-generator.utils";
+import GitHubContribution from "@/components/GitHubContribution";
+
+const MAX_ITERATIONS = 50_000;
+
+type Algorithm =
+  | "sha256"
+  | "sha512"
+  | "md5"
+  | "pbkdf2"
+  | "hmac-sha256"
+  | "hmac-sha512";
+
+interface Option {
+  value: string;
+  label: string;
+}
+
+const algorithmOptions: Option[] = [
+  { value: "sha256", label: "SHA-256" },
+  { value: "sha512", label: "SHA-512" },
+  { value: "md5", label: "MD5" },
+  { value: "pbkdf2", label: "PBKDF2 (with SHA-512)" },
+  { value: "hmac-sha256", label: "HMAC-SHA-256" },
+  { value: "hmac-sha512", label: "HMAC-SHA-512" },
+];
+
+const encodingOptions: Option[] = [
+  { value: "hex", label: "Hex" },
+  { value: "base64", label: "Base64" },
+  { value: "latin1", label: "Latin1" },
+];
+
+export default function HashGenerator() {
+  const [textInput, setTextInput] = useState("");
+  const [algorithm, setAlgorithm] = useState<Algorithm>("sha256");
+  const [saltInput, setSaltInput] = useState("");
+  const [secretKey, setSecretKey] = useState("");
+  const [iterations, setIterations] = useState(10000);
+  const [outputLength, setOutputLength] = useState(64);
+  const [encoding, setEncoding] = useState<BinaryToTextEncoding>("hex");
+  const [hashOutput, setHashOutput] = useState("");
+  const { buttonText, handleCopy } = useCopyToClipboard();
+
+  const handleTextChange = useCallback(
+    (event: React.ChangeEvent<HTMLTextAreaElement>) => {
+      setTextInput(event.currentTarget.value);
+    },
+    []
+  );
+
+  const handleSaltChange = useCallback(
+    (event: React.ChangeEvent<HTMLInputElement>) => {
+      setSaltInput(event.currentTarget.value);
+    },
+    []
+  );
+
+  const handleSecretKeyChange = useCallback(
+    (event: React.ChangeEvent<HTMLInputElement>) => {
+      setSecretKey(event.currentTarget.value);
+    },
+    []
+  );
+
+  const handleAlgorithmChange = useCallback((value: string) => {
+    setAlgorithm(value as Algorithm);
+  }, []);
+
+  const handleIterationsChange = useCallback(
+    (event: React.ChangeEvent<HTMLInputElement>) => {
+      const value = Number(event.currentTarget.value);
+      setIterations(Math.min(value, MAX_ITERATIONS));
+    },
+    []
+  );
+
+  const handleOutputLengthChange = useCallback(
+    (event: React.ChangeEvent<HTMLInputElement>) => {
+      setOutputLength(Number(event.currentTarget.value));
+    },
+    []
+  );
+
+  const handleEncodingChange = useCallback((value: string) => {
+    setEncoding(value as BinaryToTextEncoding);
+  }, []);
+
+  const hashFunctions = useMemo(() => {
```

Comment:
It might be good idea to have a strong type that is defined on top of the file and utilized here

```
type HashFunctions = {
  [K in Algorithm]: () => string;
};
```

then we can have

```
const hashFunctions: HashFunctions = useMemo()
```